### PR TITLE
[spirv] Add support for [[vk::shader_record_nv]]

### PR DESCRIPTION
**Pull request by alelenv**
@ehsannas 
Please review fix for issue #2019 
Tests have been added along with documentation update.

Thanks

### CODE REVIEW

**Comment by ehsannas** [comment]
Great! Thanks @alelenv . I'll take a look.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 1]
```diff
@@ -852,6 +856,51 @@ SpirvVariable *DeclResultIdMapper::createPushConstant(const VarDecl *decl) {
...
+  const std::string structName =
+      "type.ShaderRecordBufferNV." + decl->getName().str();
+  // The front-end does not allow arrays of cbuffer/tbuffer.
+  SpirvVariable *bufferVar = createStructOrStructArrayVarOfExplicitLayout(
+      decl, /*arraySize*/ 0, ContextUsageKind::ShaderRecordBufferNV, structName,
+      decl->getName());
+
+  // We still register all VarDecls seperately here. All the VarDecls are
+  // mapped to the <result-id> of the buffer object, which means when querying
+  // querying the <result-id> for a certain VarDecl, we need to do an extra
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Typo: 'querying' repeated ;)

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 2]
```diff
@@ -1151,6 +1151,30 @@ bool SpirvEmitter::validateVKAttributes(const NamedDecl *decl) {
...
+
+    if (!isValidType) {
+      emitError(
+          "vk::shader_record_nv can be applied only to cbuffer/ConstantBuffer",
+          loc);
+      success = false;
+    }
+    if (decl->hasAttr<VKBindingAttr>()) {
+      emitError("vk::shader_record_nv attribute cannot be used together with "
+                "vk::binding attribute",
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp]: Can we add a test to make sure this error is emitted when needed?

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 3]
```diff
@@ -852,6 +856,51 @@ SpirvVariable *DeclResultIdMapper::createPushConstant(const VarDecl *decl) {
...
+      "type.ShaderRecordBufferNV." + recordType->getDecl()->getName().str();
+  SpirvVariable *var = createStructOrStructArrayVarOfExplicitLayout(
+      recordType->getDecl(), /*arraySize*/ 0,
+      ContextUsageKind::ShaderRecordBufferNV, structName, decl->getName());
+
+  // Register the VarDecl
+  astDecls[decl] = DeclSpirvInfo(var);
+
+  // Do not push this variable into resourceVars since it does not need
+  // descriptor set.
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: I think you should also add
```
// CHECK-NOT: OpDecoration %bla DescriptorSet
// CHECK-NOT: OpDecoration %bla Binding
```
to your tests?

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp (hunk 2 above)
Comment (alelenv) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp]: Added test to vk.attribute.invalid.hlsl

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 1 above)
Comment (alelenv) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Done

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (alelenv) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Yes, this is a good idea. Added to tests

**Comment by alelenv** [comment]
Thanks for quick turnaround of review

**Pull request merged**